### PR TITLE
Policy track derived from rules better

### PR DESCRIPTION
**Pull request by jrajahalme**
MapStateEntry metadata was improperly merged when an allow entry and deny entry were merged. Fix this by only merging entries when they both are either allow or deny entries.

Add identical keys also when the entries are different regarding all other fields than just the owners. This way DerivedFromRules labels get merged also if the owning CachedSelectors happen to be the same but are used in rules with different labels.

When inserting an allow entry as a deny due to covering deny rule, DerivedFromRules was based on the allow rule, when it should be a combination of all the covering deny rules instead.

Add `MapStateEntry.DatapathAndDerivedFromEqual` to be used in testing via `mapState.Equal` and `mapState.Diff`. This will help find bugs in properly tracking the `DerivedFromRules` field, which has so far been ignored in testing when using `mapsState.Equal` and `mapState.Diff`.

Without these changes the resulting `DerivedFromRules` labels depended on the insertion order of MapState entries, causing intermittent test flakes due to essentially random iteration order of Go maps when test code was changed to care about the resulting `DerivedFromRules` labels by the 1st commit.

```release-note
Policy rule labels are tracked more accurately in endpoint policy maps.
```

### CODE REVIEW

File: pkg/policy/mapstate.go [hunk 1]
```diff
@@ -892,22 +877,13 @@ func (ms *mapState) RemoveDependent(owner Key, dependent Key, identities Identit
 
 // Merge adds owners, dependents, and DerivedFromRules from a new 'entry' to an existing
 // entry 'e'. 'entry' is not modified.
-// IsDeny, ProxyPort, and AuthType are merged by giving precedence to deny over non-deny, proxy
-// redirection over no proxy redirection, and explicit auth type over default auth type.
+// Merge is only called if both entries are allow or deny entries, so deny precedence is not
```
Comment (squeed) [inline pkg/policy/mapstate.go:889]: Can we enforce this with an assertion / panic?

File: pkg/policy/mapstate.go [hunk 2]
```diff
@@ -1349,38 +1323,35 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
...
+				insertAsDeny = true
+				denyEntry = NewMapStateEntry(k, v.DerivedFromRules, 0, "", 0, true, DefaultAuthType, AuthTypeDisabled)
+			} else {
+				// Collect the owners and labels of all the contributing deny rules
+				denyEntry.Merge(&v)
+			}
 			return true
 		})
-		if bailed || changeToDeny {
+		if bailed || insertAsDeny {
```
Comment (squeed) [inline pkg/policy/mapstate.go]: can we remove this nested `if bailed` ?

File: pkg/policy/mapstate.go [hunk 3]
```diff
@@ -1069,9 +1041,10 @@ func (ms *mapState) addKeyWithChanges(key Key, entry MapStateEntry, identities I
 		// Compare for datapath equalness before merging, as the old entry is updated in
 		// place!
 		datapathEqual = oldEntry.DatapathEqual(&entry)
+
 		oldEntry.Merge(&entry)
 		ms.insert(key, oldEntry, identities)
-	} else {
+	} else if !exists || entry.IsDeny {
```
Comment (squeed) [inline pkg/policy/mapstate.go:1063]: Can you comment why `entry.IsDeny` is special here?

**Comment by squeed** [comment]
Looks basically good. This makes the logic just a hair simpler, which I always appreciate.

AFAICT we have all the cases covered, but a few more comments would be useful.

File: pkg/policy/mapstate.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/mapstate.go:889]: No asserts or panics in production code, so the best we can do here IMO is to bail out and log a warning.

File: pkg/policy/mapstate.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/policy/mapstate.go]: done :-)

File: pkg/policy/mapstate.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/policy/mapstate.go:1063]: Added this comment:
```
	// Insert a new entry if one did not exist or a deny entry is overwriting an allow
	// entry.
```

**Review by squeed — APPROVED** [review]

File: pkg/policy/mapstate.go [hunk 4]
```diff
@@ -1162,8 +1167,10 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
 	// Sanity check on the newKey
 	if newKey.TrafficDirection >= trafficdirection.Invalid.Uint8() {
 		stacktrace := hclog.Stacktrace()
```
Comment (derailed) [inline pkg/policy/mapstate.go:1169]: nit: Don't really need this var. Might as well the make the call in the log line directly.

File: pkg/policy/mapstate.go [hunk 5]
```diff
@@ -892,22 +886,20 @@ func (ms *mapState) RemoveDependent(owner Key, dependent Key, identities Identit
...
+// considered here.
+// ProxyPort, and AuthType are merged by giving precedence to proxy redirection over no proxy
+// redirection, and explicit auth type over default auth type.
 func (e *MapStateEntry) Merge(entry *MapStateEntry) {
-	// Deny is sticky
-	if !e.IsDeny {
-		e.IsDeny = entry.IsDeny
+	// Bail out loudly if both entries are not denies or allows
+	if e.IsDeny != entry.IsDeny {
+		stacktrace := hclog.Stacktrace()
```
Comment (derailed) [inline pkg/policy/mapstate.go:896]: nit: No need for this var. Might as well call it directly in the log line below.

**Review by derailed — APPROVED** [review]
@jrajahalme Nice work!

**Pull request merged**